File: package/main/src/Validate/array/arrayOf.ts
import { isArray } from "@/Validate/isArray";
import {
  attachStandard,
  type StandardSchemaV1,
} from "@/Validate/standardSchema";
import type { ValidateCoreReturnType, ValidateType } from "@/Validate/type";

// Extract the validated value type by reading the validator's `type` field
// (and applying `ValidateType` to map type tags like "string" back to the
// runtime type). Reading the field directly lets validators that expose the
// literal union via the `type` field (such as `oneOf`) flow through arrayOf
// without being collapsed to `string`.
export type ArrayOfExtractValidatedType<V> = V extends (value: never) => {
  type: infer T;
}
  ? ValidateType<T>
  : never;

/**
 * Creates an array validator that validates every element with a single validator
 * @template V - The validator function applied to each element (its return `type` field carries the element type)
 * @param {V} validator - Validator applied to each element (e.g. an object validator)
 * @param {string} [message] - Custom error message for array type validation
 * @returns {Function} - Validator function for arrays whose elements satisfy the given validator
 */
export const arrayOf = <
  V extends (value: never) => {
    type: unknown;
    message: string;
    validate: boolean;
  },
>(
  validator: V,
  message?: string,
): ((
  values: ArrayOfExtractValidatedType<V>[],
) => ValidateCoreReturnType<ArrayOfExtractValidatedType<V>[]>) &
  StandardSchemaV1<
    ArrayOfExtractValidatedType<V>[],
    ArrayOfExtractValidatedType<V>[]
  > => {
  type Element = ArrayOfExtractValidatedType<V>;
  const arrayValidator = (
    values: Element[],
  ): ValidateCoreReturnType<Element[]> => {
    if (!isArray(values)) {
      return {
        validate: false,
        message: message ?? "",
        type: values,
      };
    }
    for (const value of values) {
      const result = (
        validator as unknown as (v: Element) => {
          validate: boolean;
          message: string;
        }
      )(value);
      if (!result.validate) {
        return {
          validate: false,
          message: result.message,
          type: values,
        };
      }
    }
    return {
      validate: true,
      message: "",
      type: values,
    };
  };
  return attachStandard<Element[], Element[], typeof arrayValidator>(
    arrayValidator,
  );
};

File: package/umt_python/src/validate/array_of.py
"""Array validation core module.

Provides a higher-order validator that applies a single element validator to
every item of a list. It is the Python counterpart of the TypeScript
``arrayOf`` factory.
"""

from collections.abc import Callable
from dataclasses import dataclass
from typing import Protocol


class ElementResult(Protocol):
    """
    Minimal shape returned by an element validator.

    Only the ``validate`` and ``message`` fields are read while iterating over
    the array, so any validator return type exposing them is accepted.

    Attributes:
        validate: Whether the element passed validation.
        message: Error message describing why the element failed.
    """

    validate: bool
    message: str


@dataclass
class ArrayOfReturnType:
    """
    Result produced by an ``array_of`` validator.

    The ``type`` field preserves the original list so downstream composition
    helpers can flow the validated value through unchanged.

    Attributes:
        validate: Whether every element passed validation.
        message: Error message describing the first failure, empty on success.
        type: The original value passed to the validator.
    """

    validate: bool
    message: str
    type: object


def array_of(
    validator: Callable[[object], ElementResult],
    message: str = "",
) -> Callable[[object], ArrayOfReturnType]:
    """
    Creates an array validator that validates every element with a single
    validator.

    Args:
        validator: Validator applied to each element (for example an object
            validator).
        message: Custom error message reported when the value is not a list
            (default: empty string).

    Returns:
        A validator function for lists whose elements satisfy the given
        validator. It stops at the first failing element and propagates that
        element's message.

    Example:
        >>> from src.validate.is_string import is_string
        >>> def string_validator(value: object) -> ArrayOfReturnType:
        ...     ok = is_string(value)
        ...     return ArrayOfReturnType(
        ...         validate=ok,
        ...         message="" if ok else "not a string",
        ...         type=value,
        ...     )
        >>> array_of(string_validator)(["a", "b"]).validate
        True
        >>> array_of(string_validator)(["a", 1]).validate
        False
        >>> array_of(string_validator)("not a list").validate
        False
    """

    def array_validator(values: object) -> ArrayOfReturnType:
        if not isinstance(values, list):
            return ArrayOfReturnType(validate=False, message=message, type=values)
        for value in values:
            result = validator(value)
            if not result.validate:
                return ArrayOfReturnType(
                    validate=False, message=result.message, type=values
                )
        return ArrayOfReturnType(validate=True, message="", type=values)

    return array_validator
